common/js/src/i18n.js

```diff
@@ -0,0 +1,100 @@
+/** @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This file contains common functionality for
+ * internationalization.
+ */
+
+
+goog.provide('GoogleSmartCard.I18n');
+
+goog.require('GoogleSmartCard.Logging');
+goog.require('goog.log.Logger');
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/**
+ * @type {!goog.log.Logger}
+ * @const
+ */
+var logger = GSC.Logging.getScopedLogger('GoogleSmartCard.I18n');
+
+/** @const */
+var I18N_DATA_ATTRIBUTE = 'data-i18n';
+
+/** @const */
+var I18N_DATA_ARIA_LABEL_ATTRIBUTE = 'data-i18n-aria-label';
+
+/**
+ * @param {!Element} element
+ * @param {string} attribute
+ * @param {function(!Element,string)} transformFunction
+ * @private
+ */
+function transformElement(element, attribute, transformFunction) {
+  var i18nId = element.getAttribute(attribute);
+  if (!i18nId) {
+    GSC.Logging.failWithLogger(
+        logger,
+        'Failed to get element attribute: ' + element.outerHTML);
+  }
+
+  var translatedText = chrome.i18n.getMessage(i18nId);
+  if (!translatedText) {
+    GSC.Logging.failWithLogger(
+        logger,
+        'Failed to get translation for text with id: ' + i18nId);
+  }
+
+  transformFunction(element, translatedText);
+}
+
+/**
+ * @param {string} attribute
+ * @param {function(!Element,string)} transformFunction
+ * @private
+ */
+function transformAllElements(attribute, transformFunction) {
+  var selector = '[' + attribute + ']';
+  for (let element of document.querySelectorAll(selector)) {
+    transformElement(element, attribute, transformFunction);
+  }
+}
+
+/**
+ * Goes through all the HTML elements in the current window, replacing
+ * element.textContent with translation for elements containing attribute
+ * I18N_DATA_ATTRIBUTE, and setting attribute aria-label to translation for
+ * elements containing attribute I18N_DATA_ARIA_LABEL_ATTRIBUTE. Used for
+ * translation that plays nice with Chromevox (accessibility tool).
+ */
+GSC.I18n.adjustElementsTranslation = function() {
+  transformAllElements(
+      I18N_DATA_ATTRIBUTE, function(element, translatedText) {
+    element.textContent = translatedText;
```

Comment:
(this is optional) Could be useful to emit log messages describing each transformation with the DOM (probably, at FINE level, so that by default they go to the console only in Debug mode).